keps/sig-auth/20190607-certificates-api.md

```diff
@@ -132,6 +133,85 @@ A typical successful issuance proceeds as follows.
+    3. Permitted x509 extensions - none
+    4. Permitted key usages - exactly `[]string{"key encipherment", "digital signature", "client auth"}`
+    5. Expiration/cert lifetime - minimum of CSR signer or request.  Sanity of the time is the concern of the signer.
+    6. CA bit allowed/disallowed - not allowed.
+ 3. kubernetes.io/kubelet-serving - signs serving certificates that are honored as a valid kubelet serving certificate 
+    by the kube-apiserver, but has no other guarantees.  Never auto-approved by kube-controller-manager.
+    1. Trust distribution: signed certificates must be honored by the kube-apiserver as valid to terminate connections to a kubelet.
+       The CA bundle is not distributed by any other means.
+    2. Permitted subjects - organizations are exactly `[]string{"system:nodes"}`, common name starts with `"system:node:"`
+    3. Permitted x509 extensions - SAN DNS has value
+    4. Permitted key usages - exactly `[]string{"key encipherment", "digital signature", "server auth"}`
+    5. Expiration/cert lifetime - minimum of CSR signer or request.
+    6. CA bit allowed/disallowed - not allowed.
+ 4. kubernetes.io/legacy-unknown - has no guarantees for trust at all.  Some distributions may honor these as client
+    certs, but that behavior is not standard kubernetes behavior.  Never auto-approved by kube-controller-manager.
+    1. Trust distribution: None.  There is no standard trust or distribution for this signer in a kubernetes cluster.
+    2. Permitted subjects - any
+    3. Permitted x509 extensions - honors SAN extensions and discards other extensions.
+    4. Permitted key usages - may include `[]string{"key encipherment", "digital signature", "client auth", "server auth"}`
+    5. Expiration/cert lifetime - minimum of CSR signer or request.  Sanity of the time is the concern of the signer.
+    6. CA bit allowed/disallowed - not allowed.
+
+Distribution of trust happens out of band for these signers.  Any trust outside of those described above are strictly
+coincidental.  For instance, some distributions may honor kubernetes.io/legacy-unknown as client-certificates for the
+kube-apiserver, but this is not a standard.
+None of these usages are related to ServiceAccount token secrets `.data[ca.crt]` in any way.  That ca-bundle is only
+guaranteed to verify a connection the kube-apiserver using the default service.
+
+To support HA upgrades, the kube-controller-manager will duplicate defaulting code for an empty `signerName` for one
+release.
+
+#### Limiting approval powers for certain signers.
+Given multiple signers which may be implemented as "dumb" controllers that sign if the CSR is approved, there is benefit
+to providing a simple way to subdivide approval powers through the API.  We can introduce an admission plugin that requires
+ 1. verb == `create`
+ 2. resource == `certificatesigningrequests/approve/(.spec.signerName)`
```

Comment:
needs to re-check the RBAC evaluator. proxy takes a long URL and it's guarded and doesn't use a `/`.  This may be improper and need to be changed to

```
resource = certificatesigners
name = <.spec.signerName>
```

This would run out subdivision based on all signers matching a pattern, but realistically matching patterns for approval doesn't seem necessary since the API only works effectively with a numerable list.

Reply:
> This would run out subdivision based on all signers matching a pattern, but realistically matching patterns for approval doesn't seem necessary since the API only works effectively with a numerable list.

fwiw consumers like cert-manager are likely to encode their various issuers into a signerName, like `issuer.cert-manager.jetstack.io/$namespace.$name`

Reply:
Yep, our 'issuer' construct (and any similar 'out of tree' issuer concept that people may build) can be seen as an 'instantiation' of a signer, meaning that new signers with e.g. different account credentials can be dynamically registered.

Without wildcard matching, every time a user creates a new issuer/signer, RBAC rules would need to be provisioned to permit consumers of that signer access to approve for it. This could be quite painful in environments where other tooling (i.e. our 'Certificates' controller) provides abstractions on top of the CSR API as 'issuers' are namespaced resources, meaning every time an Issuer is created, an administrator/someone with the knowledge of what consumers of that signer exist in the cluster, will need to appropriately configure RBAC to allow those controllers to work with them.

Reply:
I have a preference for using name here. Is there any reason to use resource?

Reply:
> I have a preference for using name here. Is there any reason to use resource?

It gave us a clean shot at creating a real subresource, but in light of the difficulty in using it, I've updated the KEP to use a the "approve" verb on a "certificatesigners" resource

Reply:
Ok, that works for me.